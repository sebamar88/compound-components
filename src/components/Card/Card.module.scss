.card {
    box-shadow: 1px 1px 5px 0px rgba(0, 0, 0, 0.75);
    border-radius: 5px;
    padding: 10px;
    margin: 10px;
    background-color: #fefefe;
}

.cardHeader,
.cardBody,
.cardFooter {
    padding: 10px;
}

.cardTitle {
    font-size: 1.5rem;
    font-weight: 600;
    margin: 0;
}

.cardImage {
    width: 100%;
    height: auto;
}

.cardText {
    font-size: 1rem;
    font-weight: 400;
    margin: 0;
}
