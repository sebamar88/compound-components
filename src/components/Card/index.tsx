import { Props } from "../types";
import { createContext, useContext } from "react";
import styles from "./Card.module.scss";

const CardContext = createContext({ card: false });
const HeaderContext = createContext({ header: false });
const BodyContext = createContext({ body: false });
const FooterContext = createContext({ footer: false });

function useCardContext() {
    const context = useContext(CardContext);

    if (!context.card) {
        return null;
    }
    return context;
}

function useHeaderContext() {
    const context = useContext(HeaderContext);

    if (!context.header) {
        return null;
    }

    return context;
}

function useTextContext() {
    const bodyContext = useContext(BodyContext);
    const footerContext = useContext(FooterContext);

    if (!bodyContext.body && !footerContext.footer) {
        return null;
    }

    return bodyContext || footerContext;
}

const Card = ({ children }: Props) => {
    return (
        <CardContext.Provider
            value={{
                card: true,
            }}
        >
            <div className={styles.card}>{children}</div>
        </CardContext.Provider>
    );
};

const Header = ({ children }: Props) => {
    const context = useCardContext();
    if (!context) return null;
    return (
        <HeaderContext.Provider
            value={{
                header: true,
            }}
        >
            <div className={styles.cardHeader}>{children}</div>
        </HeaderContext.Provider>
    );
};

const Title = ({ children }: Props) => {
    const context = useHeaderContext();
    if (!context) return null;
    return <h2 className={styles.cardTitle}>{children}</h2>;
};

type ImageProps = {
    src: string;
};

const Image = ({ src }: ImageProps) => {
    const context = useHeaderContext();
    if (!context) return null;
    return <img className={styles.cardImage} src={src} alt="Card image" />;
};

const Text = ({ children }: Props) => {
    const context = useTextContext();
    if (!context) return null;
    return <p className={styles.cardText}>{children}</p>;
};

const Body = ({ children }: Props) => {
    return (
        <BodyContext.Provider
            value={{
                body: true,
            }}
        >
            <div className={styles.cardBody}>{children}</div>
        </BodyContext.Provider>
    );
};

const Footer = ({ children }: Props) => {
    return (
        <FooterContext.Provider
            value={{
                footer: true,
            }}
        >
            <div className={styles.cardFooter}>{children}</div>
        </FooterContext.Provider>
    );
};

Card.Title = Title;
Card.Image = Image;
Card.Header = Header;
Card.Body = Body;
Card.Footer = Footer;
Card.Text = Text;

export default Card;
