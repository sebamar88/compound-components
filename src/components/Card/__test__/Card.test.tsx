/**
 * @jest-environment jsdom
 */

import Card from "../index";
import { render } from "@testing-library/react";

const cards = [
    {
        title: "Card 1",
        text: "This is card 1",
        image: "https://picsum.photos/300/200",
        footerText: "Footer Text 1",
    },
    {
        title: "Card 2",
        text: "This is card 2",
        image: "https://picsum.photos/300/200",
        footerText: "Footer Text 2",
    },
    {
        title: "Card 3",
        text: "This is card 3",
        image: "https://picsum.photos/300/200",
        footerText: "Footer Text 3",
    },
];

describe("Card Testing", () => {
    it("should render Card", () => {
        const wrapper = render(
            <div>
                {cards.map((card) => (
                    <Card>
                        <Card.Header>
                            <Card.Title>{card.title}</Card.Title>
                            <Card.Image src={card.image} />
                        </Card.Header>
                        <Card.Body>
                            <Card.Text>{card.text}</Card.Text>
                        </Card.Body>
                        <Card.Footer>
                            <Card.Text>{card.footerText}</Card.Text>
                        </Card.Footer>
                    </Card>
                ))}
            </div>
        );
        expect(wrapper).toMatchSnapshot();
    });
});
