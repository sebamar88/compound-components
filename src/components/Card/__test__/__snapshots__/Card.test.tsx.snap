// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Testing should render Card 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="card"
        >
          <div
            class="cardHeader"
          >
            <h2
              class="cardTitle"
            >
              Card 1
            </h2>
            <img
              alt="Card image"
              class="cardImage"
              src="https://picsum.photos/300/200"
            />
          </div>
          <div
            class="cardBody"
          >
            <p
              class="cardText"
            >
              This is card 1
            </p>
          </div>
          <div
            class="cardFooter"
          >
            <p
              class="cardText"
            >
              Footer Text 1
            </p>
          </div>
        </div>
        <div
          class="card"
        >
          <div
            class="cardHeader"
          >
            <h2
              class="cardTitle"
            >
              Card 2
            </h2>
            <img
              alt="Card image"
              class="cardImage"
              src="https://picsum.photos/300/200"
            />
          </div>
          <div
            class="cardBody"
          >
            <p
              class="cardText"
            >
              This is card 2
            </p>
          </div>
          <div
            class="cardFooter"
          >
            <p
              class="cardText"
            >
              Footer Text 2
            </p>
          </div>
        </div>
        <div
          class="card"
        >
          <div
            class="cardHeader"
          >
            <h2
              class="cardTitle"
            >
              Card 3
            </h2>
            <img
              alt="Card image"
              class="cardImage"
              src="https://picsum.photos/300/200"
            />
          </div>
          <div
            class="cardBody"
          >
            <p
              class="cardText"
            >
              This is card 3
            </p>
          </div>
          <div
            class="cardFooter"
          >
            <p
              class="cardText"
            >
              Footer Text 3
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="card"
      >
        <div
          class="cardHeader"
        >
          <h2
            class="cardTitle"
          >
            Card 1
          </h2>
          <img
            alt="Card image"
            class="cardImage"
            src="https://picsum.photos/300/200"
          />
        </div>
        <div
          class="cardBody"
        >
          <p
            class="cardText"
          >
            This is card 1
          </p>
        </div>
        <div
          class="cardFooter"
        >
          <p
            class="cardText"
          >
            Footer Text 1
          </p>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="cardHeader"
        >
          <h2
            class="cardTitle"
          >
            Card 2
          </h2>
          <img
            alt="Card image"
            class="cardImage"
            src="https://picsum.photos/300/200"
          />
        </div>
        <div
          class="cardBody"
        >
          <p
            class="cardText"
          >
            This is card 2
          </p>
        </div>
        <div
          class="cardFooter"
        >
          <p
            class="cardText"
          >
            Footer Text 2
          </p>
        </div>
      </div>
      <div
        class="card"
      >
        <div
          class="cardHeader"
        >
          <h2
            class="cardTitle"
          >
            Card 3
          </h2>
          <img
            alt="Card image"
            class="cardImage"
            src="https://picsum.photos/300/200"
          />
        </div>
        <div
          class="cardBody"
        >
          <p
            class="cardText"
          >
            This is card 3
          </p>
        </div>
        <div
          class="cardFooter"
        >
          <p
            class="cardText"
          >
            Footer Text 3
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
